File: backend/src/main/java/org/superfive/telemedicine/model/Doctor.java
package org.superfive.telemedicine.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import java.time.LocalDate;
import java.util.Set;

@Entity
@AttributeOverride(name = "userID", column = @Column(name = "doctorID"))
@Getter
@Setter
public class Doctor extends User {
    // Doctor's specialty
    @ManyToOne
    @JoinColumn(name = "specialtyID", referencedColumnName = "specialtyID", nullable = false)
    private Specialty specialty;

    // Controls whether account is ACTIVE or DEACTIVATED
    @NotBlank
    private String accountStatus;

    // Doctor's appointments
    // Mapped bi-directionally to Appointment
    @JsonIgnore
    @OneToMany(
            mappedBy = "doctor",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private Set<Appointment> appointments;

    // Doctor's availability
    @JsonIgnore
    @OneToMany(
            mappedBy = "doctor",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private Set<Availability> availabilities;

    public Doctor(int doctorID, String firstName, String lastName, String email, String password, LocalDate dateOfBirth,
                  Specialty specialty, String accountStatus, Set<Availability> availabilities, Set<Appointment> appointments) {
        super(doctorID, firstName, lastName, email, password, dateOfBirth);
        this.specialty = specialty;
        this.accountStatus = accountStatus;
        this.availabilities = availabilities;
        this.appointments = appointments;
    }

    public Doctor() {
        super();
    }
}

File: frontend/lib/models/doctor.dart
import 'package:frontend/models/user.dart';
import 'package:frontend/utility.dart';

class Doctor extends User {
  final String specialty;

  Doctor({
    userID,
    firstName,
    lastName,
    email,
    dateOfBirth,
    accountStatus,
    required this.specialty}) : super(
      userID: userID,
      firstName: firstName,
      lastName: lastName,
      email: email,
      dateOfBirth: dateOfBirth,
      accountStatus: accountStatus);

  factory Doctor.fromJson(Map<String, dynamic> json) {
    return Doctor(
      userID: json['userID'],
      firstName: json['firstName'],
      lastName: json['lastName'],
      email: json['email'],
      dateOfBirth: DateTime.parse(json['dateOfBirth']),
      accountStatus: json['accountStatus'],
      specialty: json['specialty']['specialtyName'],
    );
  }

  toJson(String? newPassword, int specialtyID) {
    return {
      'userID': userID,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'password': newPassword,
      'dateOfBirth': Utility.dateToStringJSON(dateOfBirth),
      'accountStatus': accountStatus,
      'specialty': {
        'specialtyID': specialtyID,
      },
    };
  }
}
